### ISSUE TITLE
client:only ui component loses its style in a persistent island

### ISSUE THREAD

**Issue by abernier** [withastro/astro#8114]
### What version of `astro` are you using?

2.10.9

### Are you using an SSR adapter? If so, which one?

None

### What package manager are you using?

npm

### What operating system are you using?

Mac, Linux

### What browser are you using?

Chrome

### Describe the Bug

I have a `transition:persist`ent island. In that island, I get a `client:only` React component.

screencast: https://storage.googleapis.com/abernier-portfolio/flyme3.mov

When I navigate to a new page:
- ✅ island is persisted (my React component not rerendered -- that's fine)
- ❌ BUT I lose the style imported from my React component

since the Player react-component is not re-rendered, and the island persistent (`<marquee>` proves it), it's weird my `import`ed style gets touched in the head...

NB: it only happens when `client:only`

### What's the expected result?

the client UI style should be untouched across navigation

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-l21t7u?file=src%2Flayouts%2FLayout.astro,src%2Fpages%2Fmoon.astro,src%2Fcomponents%2FPlayer.tsx,src%2Fpages%2Findex.astro,src%2Fcomponents%2Fplayer.css&on=stackblitz

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by erixweb** [comment]
Hi. I've been looking at the reproducible example and what I did is change from client:only to **client:load.** This has worked for me.

**Comment by abernier** [comment]
> If you want to use client:only, documentation marks you must specify the correct framework for it to work.

I've [tested with `client:only="react"`](https://stackblitz.com/edit/github-l21t7u-5t1tam?file=src%2Flayouts%2FLayout.astro%3AL29,src%2Fpages%2Fmoon.astro,src%2Fcomponents%2FPlayer.tsx,src%2Fpages%2Findex.astro,src%2Fcomponents%2Fplayer.css&on=stackblitz) but same behaviour

**Comment by martrapp** [comment]
Preliminary analysis:
There is a <style> element in the <head> of the page that contains the styling of the player.
During swap(), it is removed because it is not part of the target page.
The <style> element is not part of either the "from" or the "to" document as sent by the server. (client:only)
It is injected into the <head> by Vite's client lib (@vite/client:updateStyle()).
I suspect this happens only once for the original page, since the island is persistent.

**Comment by matthewp** [comment]
I love this demo btw 😀

**Comment by matthewp** [comment]
Noting that this doesn't happen in prod. That's because at build time client only styles are bundled and included on the page, not injected via CSS.

**Comment by matthewp** [comment]
I think the only solution that's likely to work here is keeping styles marked with `data-vite-dev-id`. We can make this code only occur in dev.

**Comment by matthewp** [comment]
PR up

**Comment by martrapp** [comment]
Workaround while we strive to find a good solution:

Note that the issue only occurs in DEV mode, not when build/preview for production.
In DEV mode, you can import the missing CSS files explicitly on the page that uses the `client:only` component.
